magic_bridge/ic/src/magic_bridge/src/dab.rs

```diff
@@ -0,0 +1,136 @@
+use std::{str::{FromStr}, cell::RefCell, collections::HashMap};
+use ic_kit::{
+  candid::{ CandidType, Deserialize},
+  Principal,
+  ic,
+};
+
+use crate::types::*;
+use crate::factory::{CreateCanisterParam};
+
+const DAB_TOKEN_ADDRESS: &str = "zs2mv-bqaaa-aaaaa-aagla-cai";
+
+
+#[derive(CandidType, Deserialize, Clone, PartialEq, Debug)]
+pub enum DetailValue {
+    True,
+    False,
+    U64(u64),
+    I64(i64),
+    Float(f64),
+    Text(String),
+    Principal(Principal),
+    #[serde(with = "serde_bytes")]
+    Slice(Vec<u8>),
+    Vec(Vec<DetailValue>),
+}
+#[derive(CandidType, Deserialize)]
+pub struct DABParams {
+    pub name: String,
+    pub description: String,
+    pub thumbnail: String,
+    pub frontend: Option<String>,
+    pub principal_id: CanisterId,
+    pub details: Vec<(String, DetailValue)>,
+}
+
+#[derive(CandidType, Debug, Deserialize)]
+pub enum OperationError {
+    NotAuthorized,
+    NonExistentItem,
+    BadParameters,
+    Unknown(String),
+}
+
+pub type DABResponse = Result<(), OperationError>;
+
+#[derive(CandidType, Deserialize, Default, Clone)]
+pub struct DABHistory {
+    pub registered_canisters: RefCell<Vec<Principal>>,
+    pub failed_canisters: RefCell<HashMap<Principal, CreateCanisterParam>>
+}
+
+impl DABHistory {
+    pub fn add_registered_canister(&mut self, canister_id: Principal) {
+        self.registered_canisters.borrow_mut().push(canister_id);
+    }
+
+    pub fn add_failed_canister(&mut self, canister_id: Principal, params: &CreateCanisterParam) {
+        self.failed_canisters.borrow_mut().insert(canister_id, params.clone());
+    }
+}
+  
+
+
+pub async fn register_canister(canister_id: Principal, params: &CreateCanisterParam) -> Result<Principal, OperationError> {
+    let ic_stored_registry = ic::get_mut::<DABHistory>();
```

Comment:
Instead of using the global mutable, let's stick to using the single thread global store we are already using for the rest of the canister.